astropy/units/quantity_helper/function_helpers.py

```diff
@@ -438,6 +440,69 @@ def _block(arrays, max_depth, result_ndim, depth=0):
         return np_core.shape_base._atleast_nd(arrays, result_ndim)
 
 
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def arange(*args, start=None, stop=None, step=None, dtype=None, like=None):
+        return arange_impl(*args, start=start, stop=stop, step=step, dtype=dtype, like=like)
+else:
+    @dispatched_function
+    def arange(*args, start=None, stop=None, step=None, dtype=None, device=None, like=None):
+        return arange_impl(*args, start=start, stop=stop, step=step, dtype=dtype, device=device, like=like)
+
+def arange_impl(*args, start=None, stop=None, step=None, dtype=None, device=None, like=None):
+    from astropy.units import Quantity
+
+    # NumPy is supposed to validate the input parameters before this dispatched
+    # function is reached. Nevertheless, we'll sprinkle a few rundundant
+    # sanity checks in the form of `assert` statements.
+    # As they are not part of the business logic, it is fine if they are
+    # compiled-away (e.g. the Python interpreter runs with -O)
+    assert like is None
+
+    # bind positional arguments to their meaningful names
+    # following the (complex) logic of np.arange
+    assert len(args) <= 4
+    match args:
+        case pos1,:
+            assert stop is None or start is None
+            if stop is None:
+                stop = pos1
+            elif start is None:
+                start = pos1
+        case start, stop, *rest:
+            match rest:
+                # rebind step if possible
+                case step, *_:
+                    pass # nothing to do
+            if start is not None and stop is None:
+                start, stop = stop, start
+
+    has_out_unit = False
```

Comment:
The loop looks a bit like `_quantities2arrays`, except it has to deal with None. One perhaps can reverse the operations here, as in
```
kwargs = {k: v for k, v in (("start", start), ("stop", stop), ("step", step)) if v is not None}
new_values = _quantities2arrays(kwargs.values)
kwargs = {k: v for k, v in zip(kwargs.keys(), new_values)}
kwargs["dtype"] = dtype
```

Reply:
this should be addressed on my next push (but I'm riding a train so I might not succeed before I get off)